spec/mauth_client_spec.rb

```diff
@@ -52,237 +66,595 @@ def (::Rails).logger
+              signed_request = client.signed(req_w_path)
+
+              # now that we've signed the request, imagine it goes to nginx where it gets percent-decoded
+              decoded_signed_request = signed_request.clone
+              decoded_signed_request.attributes_for_signing[:request_url] = CGI.unescape(decoded_signed_request.attributes_for_signing[:request_url])
+              decoded_signed_request.attributes_for_signing[:query_string] = CGI.unescape(decoded_signed_request.attributes_for_signing[:query_string])
+              expect(authenticating_mc.authentic?(decoded_signed_request)).to be_truthy
+            end
+          end
+
+          it 'considers a request signed by an app uuid unknown to mauth to be inauthentic' do
+            bad_client = MAuth::Client.new(private_key: signing_key, app_uuid: 'nope')
+            signed_request = bad_client.signed(request)
+            stubs.get("/mauth/v1/security_tokens/nope.json") { [404, {}, []] }
+            expect(authenticating_mc.authentic?(signed_request)).to be_falsey
+          end
+
+          it 'considers a request with a bad signature to be inauthentic' do
+            signed_request = client.signed(request)
+            signed_request.headers['MCC-Authentication'] = "MWS #{app_uuid}:wat"
+            expect(authenticating_mc.authentic?(signed_request)).to be_falsey
+          end
+
+          it 'considers a request that has been tampered with to be inauthentic' do
+            signed_request = client.signed(request)
+            signed_request.attributes_for_signing[:verb] = 'DELETE'
+            expect(authenticating_mc.authentic?(signed_request)).to be_falsey
+          end
+
+          it 'considers a request with many repeated query params authentic' do
+            pairs = (1..100).reduce([]) do |acc, el|
+              acc.push(['param1', el], ['param2', el])
+            end.shuffle
+
+            request = TestSignableRequest.new(
+              verb: 'PUT',
+              request_url: '/',
+              body: 'himom',
+              query_string: pairs.map { |pair| pair.join('=') }.join('&')
+            )
+            signed_request = client.signed(request)
+            expect(authenticating_mc.authentic?(signed_request)).to be_truthy
+          end
+
+          it 'considers a signed request with multi-byte UTF-8 characters in the query string to be authentic' do
+            request = TestSignableRequest.new(
+              verb: 'PUT',
+              request_url: '/',
+              body: 'himom',
+              query_string: 'prm=val&prm=𝖛𝗮ḷ&パラメータ=値&매개 변수=값&參數=值'
+            )
+            signed_request = client.signed(request)
+            expect(authenticating_mc.authentic?(signed_request)).to be_truthy
+          end
+
+          it 'considers a signed request with multi-byte UTF-8 characters in the query string to be authentic' do
+            qs = 'prm=パ&prm=개'
```

Comment:
What is this doing?

Reply:
Whoops: changed to `considers a signed request with repeated query param keys with multi-byte UTF-8 character values to be authentic` and updated request to use this query string. This was a spec to test sorting by codepoint of values if keys are the same.